corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/api.js

```diff
@@ -18,87 +18,91 @@ FormplayerFrontend.module("Menus", function (Menus, FormplayerFrontend, Backbone
                 options,
                 menus;
 
-            if (!params.preview) {
-                // Make sure the user has access to this app
-                var currentApp = FormplayerFrontend.request("appselect:getApp", params.appId);
-                if (!currentApp) {
-                    FormplayerFrontend.trigger(
-                        'showError',
-                        gettext("Permission Denied")
-                    );
-                    FormplayerFrontend.trigger('navigateHome');
-                    return;
-                }
-            }
-
-            options = {
-                success: function (parsedMenus, response) {
-                    if (response.status === 'retry') {
-                        FormplayerFrontend.trigger('retry', response, function() {
-                            var newOptionsData = JSON.stringify($.extend(true, { mustRestore: true }, JSON.parse(options.data)));
-                            menus.fetch($.extend(true, {}, options, { data: newOptionsData }));
-                        }, gettext('Waiting for server progress'));
-                    } else if (response.hasOwnProperty('exception')){
-                        FormplayerFrontend.trigger('clearProgress');
-                        FormplayerFrontend.trigger(
-                            'showError',
-                            response.exception || FormplayerFrontend.Constants.GENERIC_ERROR,
-                            response.type === 'html'
-                        );
-                        FormplayerFrontend.trigger('navigation:back');
-                    } else {
-                        FormplayerFrontend.trigger('clearProgress');
-                        defer.resolve(parsedMenus);
-                        // Only configure menu debugger if we didn't get a form entry response
-                        if (!(response.session_id)) {
-                            FormplayerFrontend.trigger('configureDebugger');
-                        }
-                    }
-                },
-                error: function (_, response) {
-                    if (response.status === 423) {
-                        FormplayerFrontend.trigger(
-                            'showError',
-                            Formplayer.Errors.LOCK_TIMEOUT_ERROR
-                        );
-                    } else {
+            $.when(FormplayerFrontend.request("appselect:apps")).done(function (appCollection) {
+                if (!params.preview) {
+                    // Make sure the user has access to the app
+                    if (!appCollection.find(function(app) {
+                        return app.id === params.appId || app.get('copy_of') === params.copyOf;
```

Comment:
Can you explain this one a little?  `appCollection` is a set of the most recent app builds, right?  Are there scenarios where `app.id === params.copyOf` or `app.copy_of === params.appId`?

I'm wondering if this should just check if there's anything in the set of all `app.id`s and `app.copy_of`s that matches either `params.appId` or `params.copyOf`.  That does sound a little heavy-handed, though.

Reply:
Yep, `appCollection` is the most recent app builds, created by [get_web_apps_available_to_user](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/cloudcare/views.py#L99) and used to render the set of app links that you see on the web apps home page. So `app.id` is always going to be a build id, and `app.copy_of` will always have the canonical app id.

`params.appId` is always going to be populated, and it'll always be a build id. For the links on the web apps home page, `params.appId` is always going to match one of the app ids in `appCollection`. For these links, `params.copyOf` doesn't need to be provided. For the links from the releases page, `params.appId` is still a build id, but it's likely not the latest build. These links now provide `params.copyOf` as a fallback.

So basically the first half of the conditional is for web apps home page links and the second half is for release page links.

Reply:
Thanks, that makes sense.  What about viewing an un-built, "current" application in webapps or live preview?  Will that have `params.appID` set to the canonical HQ app ID (ie `app.copy_of`) and `params.copyOf` set to null?

Reply:
No - you never really view an unbuilt app in web apps, even if set to view the current app (we make a build behind the scenes if one doesn't exist).